embedded/xsens_imu/src/imu.c

```diff
@@ -0,0 +1,216 @@
+/*
+* imu.c
+* Author: Rohan Daruwala
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <pthread.h>
+#include <semaphore.h> 
+#include <unistd.h>
+#include "imu.h"
+#include "i2c.h"
+
+//Global Variables
+IMU_data * data;
+i2c_settings * i2c;
+pthread_t IMUThread;
+sem_t vMutex, oMutex;
+
+void SetupIMU(){
+	sem_init(&vMutex, 0, 1);
+	sem_init(&oMutex, 0, 1);
+	i2c = malloc(sizeof(i2c_settings));
+	i2c->bus = 2;
+	i2c->deviceAddress = I2C_ADDRESS;
+	i2c->openMode = O_RDWR;
+
+	data = malloc(sizeof(IMU_data));
+
+	if (i2c_begin(i2c) == -1){
+		fprintf(stderr, "Could not open i2c bus.\n");
+		return;
+	}
+
+	if (pthread_create(&IMUThread, NULL, IMULoop, NULL)){
+		fprintf(stderr, "Error creating IMU thread\n");
+	}
+
+	//Set all IMU structure values to zero
+	data->dVx = 0;
+	data->dVy = 0;
+	data->dVz = 0;
+
+	data->accelX = 0;
+	data->accelY = 0;
+	data->accelZ = 0;
+
+	data->q0 = 0;
+	data->q1 = 0;
+	data->q2 = 0;
+	data->q3 = 0;
+}
+
+void *IMULoop(){
+
```

Comment:
Can we declare variables in the `void *IMULoop` scope instead of the `while (1)` scope so we know for sure we're not constantly moving the stack pointer to create the local variables for each iteration?

I don't think GCC can perform that optimization otherwise